File: 974/974.kt
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/974.py

val DIGITS_LIST = listOf(1, 3, 5, 7, 9)
val BIT_TABLE = mapOf(1 to 0, 3 to 1, 5 to 2, 7 to 3, 9 to 4)
val ALL_MASK = (1 shl 5) - 1

fun computeCountLen(bigL: Int): Long {
    val memo = HashMap<Long, Long>()
    fun encode(pos: Int, m3: Int, m7: Int, msk: Int): Long =
        pos.toLong() * (3L * 7L * 32L) + m3.toLong() * (7L * 32L) + m7.toLong() * 32L + msk.toLong()
    fun dp(pos: Int, mod3: Int, mod7: Int, mask: Int): Long {
        if (pos == bigL) {
            return if (mod3 == 0 && mod7 == 0 && mask == ALL_MASK) 1L else 0L
        }
        val key = encode(pos, mod3, mod7, mask)
        memo[key]?.let { return it }
        val choices = if (pos == bigL - 1) listOf(5) else DIGITS_LIST
        var total = 0L
        for (d in choices) {
            total += dp(pos + 1, (mod3 * 10 + d) % 3, (mod7 * 10 + d) % 7, mask xor (1 shl BIT_TABLE[d]!!))
        }
        memo[key] = total
        return total
    }
    return dp(0, 0, 0, 0)
}

fun performUnrank(bigL: Int, kVal: Long): String {
    val memo = HashMap<Long, Long>()
    fun encode(pos: Int, m3: Int, m7: Int, msk: Int): Long =
        pos.toLong() * (3L * 7L * 32L) + m3.toLong() * (7L * 32L) + m7.toLong() * 32L + msk.toLong()
    fun dp(pos: Int, mod3: Int, mod7: Int, mask: Int): Long {
        if (pos == bigL) {
            return if (mod3 == 0 && mod7 == 0 && mask == ALL_MASK) 1L else 0L
        }
        val key = encode(pos, mod3, mod7, mask)
        memo[key]?.let { return it }
        val choices = if (pos == bigL - 1) listOf(5) else DIGITS_LIST
        var total = 0L
        for (d in choices) {
            total += dp(pos + 1, (mod3 * 10 + d) % 3, (mod7 * 10 + d) % 7, mask xor (1 shl BIT_TABLE[d]!!))
        }
        memo[key] = total
        return total
    }
    var mod3 = 0
    var mod7 = 0
    var mask = 0
    val out = StringBuilder()
    var remaining = kVal
    for (pos in 0 until bigL) {
        val choices = if (pos == bigL - 1) listOf(5) else DIGITS_LIST
        var found = false
        for (d in choices) {
            val cnt = dp(pos + 1, (mod3 * 10 + d) % 3, (mod7 * 10 + d) % 7, mask xor (1 shl BIT_TABLE[d]!!))
            if (remaining > cnt) {
                remaining -= cnt
            } else {
                out.append(d.toString())
                mod3 = (mod3 * 10 + d) % 3
                mod7 = (mod7 * 10 + d) % 7
                mask = mask xor (1 shl BIT_TABLE[d]!!)
                found = true
                break
            }
        }
        if (!found) throw RuntimeException("RuntimeError")
    }
    return out.toString()
}

fun computeTheta(n: Long, maxL: Int = 200): String {
    var cum = 0L
    var L = 1
    while (L <= maxL) {
        val c = computeCountLen(L)
        if (cum + c >= n) {
            return performUnrank(L, n - cum)
        }
        cum += c
        L += 2
    }
    throw Exception("ValueError")
}

fun main() {
    val r1 = computeTheta(1L, 40)
    check(r1 == "1117935") { "assert failed: $r1" }
    val r2 = computeTheta(1000L, 40)
    check(r2 == "11137955115") { "assert failed: $r2" }
    println(computeTheta(10_000_000_000_000_000L, 200))
}

File: 974/974.cs
// Adapted from https://raw.githubusercontent.com/cirosantilli/project-euler-solutions/refs/heads/master/solvers/974.py

using System;
using System.Collections.Generic;
using System.Text;

class Program974
{
    static readonly int[] DIGITS_ARR = { 1, 3, 5, 7, 9 };
    static readonly Dictionary<int, int> BIT_TABLE = new Dictionary<int, int>
    {
        { 1, 0 }, { 3, 1 }, { 5, 2 }, { 7, 3 }, { 9, 4 }
    };
    static readonly int ALL_MASK = (1 << 5) - 1;

    static long ComputeCountLen(int bigL)
    {
        var memo = new Dictionary<long, long>();
        long Encode(int pos, int m3, int m7, int msk) =>
            (long)pos * (3L * 7L * 32L) + (long)m3 * (7L * 32L) + (long)m7 * 32L + msk;
        long Dp(int pos, int mod3, int mod7, int mask)
        {
            if (pos == bigL)
                return (mod3 == 0 && mod7 == 0 && mask == ALL_MASK) ? 1L : 0L;
            long key = Encode(pos, mod3, mod7, mask);
            if (memo.TryGetValue(key, out long cached)) return cached;
            int[] choices = (pos == bigL - 1) ? new[] { 5 } : DIGITS_ARR;
            long total = 0L;
            foreach (int d in choices)
                total += Dp(pos + 1, (mod3 * 10 + d) % 3, (mod7 * 10 + d) % 7, mask ^ (1 << BIT_TABLE[d]));
            memo[key] = total;
            return total;
        }
        return Dp(0, 0, 0, 0);
    }

    static string PerformUnrank(int bigL, long kVal)
    {
        var memo = new Dictionary<long, long>();
        long Encode(int pos, int m3, int m7, int msk) =>
            (long)pos * (3L * 7L * 32L) + (long)m3 * (7L * 32L) + (long)m7 * 32L + msk;
        long Dp(int pos, int mod3, int mod7, int mask)
        {
            if (pos == bigL)
                return (mod3 == 0 && mod7 == 0 && mask == ALL_MASK) ? 1L : 0L;
            long key = Encode(pos, mod3, mod7, mask);
            if (memo.TryGetValue(key, out long cached)) return cached;
            int[] choices = (pos == bigL - 1) ? new[] { 5 } : DIGITS_ARR;
            long total = 0L;
            foreach (int d in choices)
                total += Dp(pos + 1, (mod3 * 10 + d) % 3, (mod7 * 10 + d) % 7, mask ^ (1 << BIT_TABLE[d]));
            memo[key] = total;
            return total;
        }
        int mod3 = 0, mod7 = 0, mask = 0;
        var outSb = new StringBuilder();
        long remaining = kVal;
        for (int pos = 0; pos < bigL; pos++)
        {
            int[] choices = (pos == bigL - 1) ? new[] { 5 } : DIGITS_ARR;
            bool found = false;
            foreach (int d in choices)
            {
                long cnt = Dp(pos + 1, (mod3 * 10 + d) % 3, (mod7 * 10 + d) % 7, mask ^ (1 << BIT_TABLE[d]));
                if (remaining > cnt)
                {
                    remaining -= cnt;
                }
                else
                {
                    outSb.Append(d.ToString());
                    mod3 = (mod3 * 10 + d) % 3;
                    mod7 = (mod7 * 10 + d) % 7;
                    mask ^= (1 << BIT_TABLE[d]);
                    found = true;
                    break;
                }
            }
            if (!found) throw new Exception("RuntimeError");
        }
        return outSb.ToString();
    }

    static string ComputeTheta(long n, int maxL = 200)
    {
        long cum = 0L;
        for (int L = 1; L <= maxL; L += 2)
        {
            long c = ComputeCountLen(L);
            if (cum + c >= n)
                return PerformUnrank(L, n - cum);
            cum += c;
        }
        throw new Exception("ValueError");
    }

    static void Main(string[] args)
    {
        string r1 = ComputeTheta(1L, 40);
        if (r1 != "1117935") throw new Exception($"assert failed: {r1}");
        string r2 = ComputeTheta(1000L, 40);
        if (r2 != "11137955115") throw new Exception($"assert failed: {r2}");
        Console.WriteLine(ComputeTheta(10_000_000_000_000_000L, 200));
    }
}
